src/structure.cpp

```diff
@@ -123,22 +169,42 @@ void structure::choose_chunkdivision(const grid_volume &thegv, int desired_num_c
   // Next, add effort volumes for PML boundary regions:
   br.apply(this);
 
+  std::vector<int> prime_factors = get_prime_factors(desired_num_chunks);
+
+  // We may have to use a different number of chunks than the user requested
+  int adjusted_num_chunks = 1;
+  for (size_t i = 0, stop = prime_factors.size(); i < stop; ++i) {
+    adjusted_num_chunks *= prime_factors[i];
+  }
+
   // Finally, create the chunks:
   num_chunks = 0;
-  chunks = new structure_chunk_ptr[desired_num_chunks * num_effort_volumes];
-  for (int i = 0; i < desired_num_chunks; i++) {
-    const int proc = i * count_processors() / desired_num_chunks;
-    grid_volume vi =
-        gv.split_by_effort(desired_num_chunks, i, num_effort_volumes, effort_volumes, effort);
-    for (int j = 0; j < num_effort_volumes; j++) {
+  chunks = new structure_chunk_ptr[adjusted_num_chunks * num_effort_volumes];
+  std::vector<grid_volume> chunk_volumes;
+
+  if (meep_geom::fragment_stats::resolution == 0 ||
+      meep_geom::fragment_stats::has_non_medium_material() || gv.dim == Dcyl ||
+      meep_geom::fragment_stats::split_chunks_evenly) {
+    for (int i = 0; i < adjusted_num_chunks; i++) {
+      grid_volume vi = gv.split_by_effort(adjusted_num_chunks, i, num_effort_volumes, effort_volumes,
+                                          effort);
+      chunk_volumes.push_back(vi);
+    }
```

Comment:
Seems like we should have continued to use `desired_num_chunks` for the `split_by_effort` code, rather than `adjusted_num_chunks` (used for the new `split_by_cost` code)?